libphal/phal_dump.C

```diff
@@ -0,0 +1,299 @@
+		    pdbg_target_path(proc), fapiRc);
+		throw dumpError_t(exception::HWP_EXECUTION_FAILURE);
+	}
+
+	log(level::INFO, "SBE RC=0x%08X", recovAction);
+	return proc;
+}
+
+/**
+ * @brief Write the dump file to the path
+ * @param[in] data Pointer to the data buffer
+ * @param[in] len Length of the buffer
+ * @param[in] dumpPath Path of the file
+ *
+ * Throws FILE_OPERATION_FAILURE in the case of error
+ */
+void writeDumpFile(char* data, size_t len, std::filesystem::path& dumpPath)
+{
+	std::ofstream outfile{dumpPath, std::ios::out | std::ios::binary};
+	if (!outfile.good()) {
+		int err = errno;
+		log(level::ERROR,
+		    "Failed to open the dump file for writing err=%d", err);
+		throw dumpError_t(exception::FILE_OPERATION_FAILURE);
+	}
+	outfile.exceptions(std::ifstream::failbit | std::ifstream::badbit |
+			   std::ifstream::eofbit);
+	try {
+		outfile.write(data, len);
+	} catch (std::ofstream::failure& oe) {
+		int err = errno;
+		log(level::ERROR,
+		    "Failed to write to dump file err=%d error message=%s", err,
+		    oe.what());
+		throw dumpError_t(exception::FILE_OPERATION_FAILURE);
+	}
+	outfile.close();
+}
+
+void collectSBEDump(uint32_t id, uint32_t failingUnit,
+		    std::filesystem::path& dumpPath)
+{
+	struct pdbg_target* proc = NULL;
+
+	log(level::INFO,
+	    "Collecting SBE dump: path=%s, id=%d, chip position=%d",
+	    dumpPath.string().c_str(), id, failingUnit);
+
+	std::stringstream ss;
+	ss << std::setw(8) << std::setfill('0') << id;
+	std::string idStr = ss.str();
+
+	std::string baseFilename =
+	    idStr + ".0_" + std::to_string(failingUnit) + "_SbeData_p10_";
+	proc = preCollection(failingUnit);
+	fapi2::ReturnCode fapiRc;
+
+	// Collect SBE local register dump
+	std::vector<SBESCOMRegValue_t> sbeScomRegValue;
+	fapiRc = p10_sbe_localreg_dump(proc, true, sbeScomRegValue);
+	if (fapiRc != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR,
+		    "Failed in p10_sbe_localreg_dump for proc=%s, rc=0x%08X",
+		    pdbg_target_path(proc), fapiRc);
+	} else {
+
+		std::string dumpFilename =
+		    baseFilename + "p10_sbe_localreg_dump";
+		std::filesystem::path basePath = dumpPath / dumpFilename;
+
+		try {
+			writeDumpFile(
+			    reinterpret_cast<char*>(&sbeScomRegValue[0]),
+			    sizeof(SBESCOMRegValue_t) * sbeScomRegValue.size(),
+			    basePath);
+		} catch (dumpError_t& e) {
+			log(level::ERROR,
+			    "Error in writing p10_sbe_localreg_dump file "
+			    "errorMsg=%s",
+			    e.what());
+		}
+	}
+
+	// Dump contents of various PIB Masters and Slaves internal registers
+	std::vector<sRegV> pibmsRegSet;
+	std::vector<sReg> regs2Dump;
+	if (pibms_regs_2dump.size()) {
+		regs2Dump.insert(regs2Dump.end(), pibms_regs_2dump.begin(),
+				 pibms_regs_2dump.end());
+	}
+	if (regs2Dump.size()) {
+		sRegV regv;
+		for (auto& reg : regs2Dump) {
+			regv.reg.attr = reg.attr;
+			regv.reg.addr = reg.addr;
+			regv.reg.name = reg.name;
+			pibmsRegSet.push_back(regv);
```

Comment:
here for sRegV - constructor is called and in push_back - copy constructor is called so 2 constructors
if we change to
for(auto& reg : regs2Dump)
{
  sRegv regv;
  pibmsRegSet.emplace_back(regv)
 }
 
 only 1 construcot for sRegv is called and a move